### ISSUE TITLE
We always use /Zi, can clcache be adapted to copy with this?

### ISSUE THREAD

**Issue by broesdecat** [frerich/clcache#30]
Hi all,

In our company, /Zi is enabled by default for our development builds.
Is there a way clcache (and/or our workflow) can be adapted so we can use it, or has any of the existing forks already done this?

Thanks a lot!
Broes

**Comment by frerich** [comment]
I think that technically it's perfectly possible. As far as I know, the only requirement is that clcache copies the generated {{.pdb}} files into/out of the cache. In fact, there's a source code comment which says the same: https://github.com/frerich/clcache/blob/master/clcache.py#L642

**Comment by inorton** [comment]
As far as I know it isn't possible, partly because pdb files are like databases. As I understand it, if you compile ten objects with the same pdb target then the pdb will only be good if none of the ten objects are rebuilt. If you rebuilt only one object, the pdb would be updated again and would be invalid for the other unaffected objects.

The best solution I had was to force Zy until the target required was a dll or exe and use Zi then. The effect is that you get the pdb you want and get caching at the object level.

**Comment by frerich** [comment]
Not sure I understand @inorton : you can have one .pdb file per object by using the '/Fd' switch (https://msdn.microsoft.com/en-us/library/9wst99a9.aspx) no? I use that in some projects, having every single 'cl' invocation generate one .obj and one .pdb file. I was thinking that this could be used for the cache.

**Comment by webmaster128** [comment]
@frerich Hot stuff. By adding `/Zi /Fd$@.pdb` globally to compile flags, one can easily set this up. You just have to replace nmake with [jom](https://wiki.qt.io/Jom), because nmake does not like the `$@` shortcut in batch mode.

So what you describe looks very doable at least for qmake projects.

**Comment by inorton** [comment]
Hiya, sorry, I didn't mean to say it was impossible. Rather that if the pdb file exists already on disk the compiler will merge the new pdb data into the file so you can only cache if the pdb file doesn't exist yet and only score hit if it is absent

**Comment by tfar** [comment]
I a lot merged PRs related to this issue. What's the current state? We also sue /Zi on our windows builds and it the latest release no cache seems to be created.

**Comment by frerich** [comment]
@tfar A lot of work was done on making cache accesses work better (and analysing the command line such that we notice that external `.pdb` files are generated in the first place...). `/Zi` itself is still unsupported, but I think we're in a better position to implement support for it than ever before. Hence this issue is still open.

**Comment by TiloW** [comment]
This option is particularly interesting since it would allow us to use `/Debug:FASTLINK`.

**Comment by inorton** [comment]
Hello,  I thought I'd update this with my own research from cclash.  

PDB files are generated by a different process (mspdbsrv). They arrive or are updated on disk after cl completes a compilation or linking operation.

One huge problem with caching them is that the pdb files are input files aswell as outputs. mspdbsrv updates the file with new debug information if the file exists beforehand.  If there are several compilations going on at once targetting the same pdb then the order the pdb gets updated is unpredictable.

All this makes caching very hard.

On the other hand, comping with /Z7 is cachable, you could then alter your final link step to use /Zi and still get the pdb you want.

**Comment by johnwbyrd** [comment]
Regarding the mspdbsrv issue: Microsoft has already apparently had problems like the one you describe, and has implemented a [workaround](https://msdn.microsoft.com/en-us/library/dn502518.aspx).  tl:dr; use /FS.

**Comment by johnwbyrd** [comment]
With regard to the mspdbsrv issue above: mspdbsrv can be bent to your will by manually starting and stopping the process yourself, and/or by using the _MSPDBSRV_ENDPOINT_ undocumented environment variable to control its behavior.

http://blog.peter-b.co.uk/2017/02/stop-mspdbsrv-from-breaking-ci-build.html

**Comment by johnwbyrd** [comment]
Oh look, someone inside Microsoft is trying to make us happy:

https://github.com/Microsoft/microsoft-pdb

**Comment by lfrankel** [comment]
One of my program's libraries fails to link in /Z7 mode, so /Zi is absolutely essential for me.

(Well, assuming #308 gets resolved, since that's stopping me from using clcache anyway).

**Comment by lfrankel** [comment]
@inorton No, that wouldn't solve this problem, the library itself will not build in /Z7 mode. Still, I'll take a look at cclash since #308 also renders clcache useless for me.

**Comment by jsmouret** [comment]
Are you shipping a library (.lib) with the pdb alongside?
Because for a dll or exe, a PDB can still be generated even when individual objs are compiled with /Z7.

**Comment by inorton** [comment]
Are you building using visual studio, msbuild or some other thing like ninja?

Is the library one giant c file or lots of small ones?

**Comment by inorton** [comment]
Oh wait. Are you saying the library doesn't link at all with /Z7 or just when using clcache?

**Comment by lfrankel** [comment]
>Are you building using visual studio, msbuild or some other thing like ninja?

Visual Studio (so, MSBuild).

> Is the library one giant c file or lots of small ones?

Lots of source files. Lots and lots and lots of source files.

> Oh wait. Are you saying the library doesn't link at all with /Z7 or just when using clcache?

Well, I haven't actually tried /Z7 without clcache, but my assumption was that yeah, it's not clcache at fault. I'll experiment when I'm back at work.

**Comment by lfrankel** [comment]
Can't reproduce the problem. Never mind.

**Comment by apeltauer** [comment]
Hi all,
any news about this? Seams that using /Z7 is still not working?!?

**Comment by izmmisha** [comment]
Hi @apeltauer,

There are few cases where PDB works in different way:
1) using PCH
2) using no PCH

If you are using PCH then you need same PDB file for obj file compiled and PCH file used. Otherwise compilation fails. But as clcache doesn't support PCH I'll omit this.

If there is no PCH, then it is possible to generate unique PDB file for each obj file.
But each PDB file will contain a lot of debug information, which are duplicates other PDB files. (same as /Z7)
On link stage all of PDBs would be merged into single file and it will take some time.

So there is no difference between /Z7 and PDB per obj.
